### PR TITLE
ft/Engineering Blog: Render a blog

### LINKED ISSUE #848: @engineeringblog - add layout for blog articles
Add a custom layout for blog posts that may contain elements like:

- [ ] Hero image
- [ ] Author metadata
- [ ] Reading time
- [ ] Category
- [ ] etc

### PR DESCRIPTION
**Pull request by kelvinkipruto**
## Description
This PR enables the rendering of blogs in the engineering blog

Fixes #848 #849

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Screenshots
image
image
image
image
image

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (koechkevin) [inline apps/engineeringblog/components/Markdown/Markdown.tsx:22]: @kelvinkipruto Are these the only languages we'll support in syntax highlighter? no .py?

Comment (m453h) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:25]: @kelvinkipruto  we should also consider `YAML` , I think that would cover all the languages we actively use

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (m453h) [inline apps/engineeringblog/components/Markdown/Markdown.tsx:17]: Since we are having both dark and light mode, should we consider making the syntax highlighter switchable maybe between `oneDark` and `oneLight` ?

**Comment by DavidTheProgrammer** [comment]
> @koechkevin @DavidTheProgrammer @m453h @thepsalmist @VinneyJ What do you think of @kilemensi comment above, to make it have less images?
> 
> > Will let others look at the code @kelvinkipruto but my vote is on a blog design that uses image sparingly i.e. only when it can improve on words. [React](https://react.dev/blog), [Next.js](https://nextjs.org/blog), etc.

I think the image quantity is fine. I like I think it looks really nice 🚀 . Looking through the rest of the codebase at the moment.

**Review by m453h — COMMENTED** [review]
We might need to add a copy to clipboard button for the syntax highlighter like this one below:


Screenshot 2024-09-03 at 16 57 37

**Review by VinneyJ — APPROVED** [review]
LGTM

**Review by VinneyJ — COMMENTED** [review]
The detail header image looks too large. What do you think of image width being the same size as the centered content? Otherwise, the code looks good to me.

**Review by DavidTheProgrammer — COMMENTED** [review]
Looks good. Just wondering if we decided against using contentlayer and went for the manual approach

File: apps/engineeringblog/utils/index.ts
Comment (kelvinkipruto) [reply apps/engineeringblog/utils/index.ts:16]: @DavidTheProgrammer ContentLayer is deprecated and [does not support NextjS 14](https://github.com/gsoft-inc/wl-hopper/issues/164) anymore. Rather than use another package, I decided to use the manual approach until we have a reason not to.

Comment (DavidTheProgrammer) [reply apps/engineeringblog/utils/index.ts:16]: Oh, is that so? Unfortunate. The manual approach is okay because we at least have gray-matter to parse the front-matter, which is key. 🚀 LGTM

**Comment by m453h** [comment]
> @koechkevin @DavidTheProgrammer @m453h @thepsalmist @VinneyJ What do you think of @kilemensi comment above, to make it have less images?
> 
> > Will let others look at the code @kelvinkipruto but my vote is on a blog design that uses image sparingly i.e. only when it can improve on words. [React](https://react.dev/blog), [Next.js](https://nextjs.org/blog), etc.

Putting an image with each post is alright, but form my observation as @VinneyJ  pointed out, the header image when you open  a post is too large and might force us to use high-res images (which increases the size of the page), we could take inspiration  from [Github Blog](https://github.blog/) though we need to think of the standard of images we will be using for each post. We would really need to emphasize on consistency of the images we select. Something like [Vercel Blog](https://vercel.com/blog)

===
Something to take note is that a minimalistic approach of no images saves us the headache of needing to select an appropriate image for each post  and means we only get to focus on writing content.

**Comment by kilemensi** [comment]
>Something to take note is that a minimalistic approach of no images saves us the headache of needing to select an appropriate image for each post and means we only get to focus on writing content.
>

Finally there!!! Use an image when/if and only when/if it improves the post **_and_** it should be part of the **content** and not _decoration_.

But, I'll let y'all decide.

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:17]: I agree @m453h. Let's do so when we are adding the light/dark mode.

Comment (kelvinkipruto) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:17]: I have added #871 to keep track of it.

**Comment by kelvinkipruto** [comment]
> Something to take note is that a minimalistic approach of no images saves us the headache of needing to select an appropriate image for each post and means we only get to focus on writing content.

Let me remove the image from the single blog page

File: apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx
Comment (kilemensi) [inline apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx:2]: For consistency, shouldn't we be using the icons we have from CfA's Figma file?

File: apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx
Comment (kilemensi) [inline apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx:6]:
```suggestion
export default function CodeCopyButton({ children }: { children: JSX.Element }) {
```

File: apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx
Comment (kilemensi) [inline apps/engineeringblog/components/Markdown/CodeCopyBtn.tsx:21]: Any reason this isn't a `Button` or `IconButton`?

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (kilemensi) [inline apps/engineeringblog/components/Markdown/Markdown.tsx:4]: Any reason this app and `RoboShield` are using different code components/functionality?

File: apps/engineeringblog/utils/index.ts
Comment (kilemensi) [inline apps/engineeringblog/utils/index.ts:16]: Not all contents will be articles/blogs, some would be just normal pages e.g. about. We need to name these functions accordingly.

**Review by kilemensi — COMMENTED** [review]
👍🏽 

---

This is last sprint's work. Can we get it over the line so folks can start working on this sprint's tasks?

**Review by koechkevin — APPROVED** [review]

File: apps/engineeringblog/app/page.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/app/page.tsx:14]: @kilemensi Because that page is server-rendered. Using any component from `commons-ui` in any server-rendered page or component throws an error about client-side components being used server-side. The common issue I have encountered is with `RichTypography` which has a `useEffect` which can only be used client-side. It is not used in `Section` but affects all `commons-ui` components if used in server-rendered components.

File: apps/engineeringblog/components/Article/ArticleCard.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/Article/ArticleCard.tsx:36]: @kilemensi Because for this app we've used the `app router` directory structure which has [a new way to mount ](https://nextjs.org/docs/messages/next-router-not-mounted#:~:text=This%20can%20also%20happen%20when%20you%20try%20to%20use%20the%20useRouter%20hook%20from%20next/router%20inside%20the%20app%20directory%2C%20as%20the%20App%20Router%27s%20useRouter%20from%20next/navigation%20has%20different%20behavior%20to%20the%20useRouter%20hook%20in%20pages.)the `router`. Using it throws an error about the router not being mounted yet.

File: apps/engineeringblog/components/Empty/Empty.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/Empty/Empty.tsx:7]: Let me rename it to `NoPosts`

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:4]: @kilemensi Yes. For `Roboshield` we had only the code. In this case, the `code` is embedded inside `Markdown`. We need to tell the `ReactMarkdown` how to render code.

**Review by m453h — APPROVED** [review]
Looks good

File: apps/engineeringblog/components/Article/ArticleCard.tsx
Comment (kilemensi) [reply apps/engineeringblog/components/Article/ArticleCard.tsx]: Cool, but when we come across errors, aren't we supposed to fix 'em?

My issue with `a` is the app will fully reload every time a user click a link.

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (kilemensi) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:4]: Cool... but my question goes both ways i.e. can the component used here also work `RoboShield`? Remember we have to maintain this code so the fewer external dependencies, the better for us.

Comment (kelvinkipruto) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:4]: No @kilemensi it can't. For `Roboshield` we need to have an editor that users can edit the code if they need to. For this, we need to format how it is displayed.

File: apps/engineeringblog/components/Article/ArticleCard.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/Article/ArticleCard.tsx]: I have used `next/link`. I will update it once #877 is merged.

File: apps/engineeringblog/app/page.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/app/page.tsx:14]: @kilemensi I have added #878 to track this.

File: apps/engineeringblog/components/Markdown/Markdown.tsx
Comment (kilemensi) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:4]: 🤔 ... Assuming you're using react-syntax-highlighter for this, isn't this their [demo](https://react-syntax-highlighter.github.io/react-syntax-highlighter/demo/prism.html) showing code editing?

... anyways, lets create an issue to revisit this. Seems like we should be able to find a single package we can use and standardise on rather than pulling in a new package everytime.

Comment (kelvinkipruto) [reply apps/engineeringblog/components/Markdown/Markdown.tsx:4]: Created #879 to handle this

File: apps/engineeringblog/components/NoPosts/NoPosts.tsx
Comment (kelvinkipruto) [reply apps/engineeringblog/components/NoPosts/NoPosts.tsx:7]: @kilemensi No. It is if there are no blogs yet rather than displaying an empty page

File: apps/engineeringblog/content/mastering-docker.mdx
Comment (kelvinkipruto) [reply apps/engineeringblog/content/mastering-docker.mdx:1]: Just added [`eslint-mdx`](https://github.com/mdx-js/eslint-mdx)

**Pull request merged**
